x/auth/middleware/block_gas.go

```diff
@@ -0,0 +1,52 @@
+package middleware
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+)
+
+type consumeBlockGasHandler struct {
+	next tx.Handler
+}
+
+// ConsumeBlockGasMiddleware check and consume block gas meter.
+func ConsumeBlockGasMiddleware(txh tx.Handler) tx.Handler {
+	return consumeBlockGasHandler{next: txh}
+}
+
+var _ tx.Handler = consumeBlockGasHandler{}
+
+func (cbgh consumeBlockGasHandler) CheckTx(ctx context.Context, req tx.Request, checkReq tx.RequestCheckTx) (res tx.Response, resCheckTx tx.ResponseCheckTx, err error) {
+	return cbgh.next.CheckTx(ctx, req, checkReq)
+}
+
+func (cbgh consumeBlockGasHandler) DeliverTx(ctx context.Context, req tx.Request) (res tx.Response, err error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	// only run the tx if there is block gas remaining
+	if sdkCtx.BlockGasMeter().IsOutOfGas() {
+		err = sdkerrors.Wrap(sdkerrors.ErrOutOfGas, "no block gas left to run tx")
+		return
+	}
+
+	startingGas := sdkCtx.BlockGasMeter().GasConsumed()
+	// If BlockGasMeter() panics it will be caught by the `RecoveryTxMiddleware` and will
+	// return an error - in any case BlockGasMeter will consume gas past the limit.
+	defer func() {
+		sdkCtx.BlockGasMeter().ConsumeGas(
+			sdkCtx.GasMeter().GasConsumedToLimit(), "block gas meter",
+		)
+
+		if sdkCtx.BlockGasMeter().GasConsumed() < startingGas {
+			panic(sdk.ErrorGasOverflow{Descriptor: "tx gas summation"})
```

Comment:
- can you add more context to the descriptor and values? (ie. gas consumed was expected to be ≥ the starting gas)
- Logging an `Error` would also be great here

Reply:
this only happens if the gas consumption overflows `uint64`, it should be very rare if not impossible, and the error `ErrorGasOverflow` seems pretty clear for this?